nipap/nipap/smart_parsing.py

```diff
@@ -0,0 +1,744 @@
+import logging
+import re
+
+import IPy
+
+from pyparsing import Combine, Forward, Group, Literal, nestedExpr, OneOrMore, ParseResults, quotedString, Regex, QuotedString, Word, ZeroOrMore, alphanums, nums, oneOf
+
+from errors import *
+
+
+class SmartParser:
+    attributes = {}
+    match_operators = ['=', '<', '>', '<=', '>=', '~']
+    boolean_operators = ['and', 'or']
+
+    def __init__(self):
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+
+    def _is_ipv4(self, ip):
+        """ Return true if given arg is a valid IPv4 address
+        """
+        try:
+            p = IPy.IP(ip)
+        except ValueError:
+            return False
+
+        if p.version() == 4:
+            return True
+        return False
+
+
+    def _is_ipv6(self, ip):
+        """ Return true if given arg is a valid IPv6 address
+        """
+        try:
+            p = IPy.IP(ip)
+        except ValueError:
+            return False
+
+        if p.version() == 6:
+            return True
+        return False
+
+
+    def _get_afi(self, ip):
+        """ Return address-family (4 or 6) for IP or None if invalid address
+        """
+
+        parts = unicode(ip).split("/")
+        if len(parts) == 1:
+            # just an address
+            if self._is_ipv4(ip):
+                return 4
+            elif self._is_ipv6(ip):
+                return 6
+            else:
+                return None
+        elif len(parts) == 2:
+            # a prefix!
+            try:
+                pl = int(parts[1])
+            except ValueError:
+                # if casting parts[1] to int failes, this is not a prefix..
+                return None
+
+            if self._is_ipv4(parts[0]):
+                if pl >= 0 and pl <= 32:
+                    # prefix mask must be between 0 and 32
+                    return 4
+                # otherwise error
+                return None
+            elif self._is_ipv6(parts[0]):
+                if pl >= 0 and pl <= 128:
+                    # prefix mask must be between 0 and 128
+                    return 6
+                # otherwise error
+                return None
+            else:
+                return None
+        else:
+            # more than two parts.. this is neither an address or a prefix
+            return None
+
+
+    def _string_to_ast(self, input_string):
+        """ Parse a smart search string and return it in an AST like form
+        """
+
+        # simple words
+        comp_word = Word(alphanums + "-./_")
+        word = Word(alphanums + "-./_").setResultsName('word')
+        # numbers
+        comp_number = Word(nums)
+        number = Word(nums).setResultsName('number')
+
+        # IPv4 address
+        ipv4_oct = Regex("((2(5[0-5]|[0-4][0-9])|[01]?[0-9][0-9]?))")
+        comp_ipv4_address = Combine(ipv4_oct + ('.' + ipv4_oct*3))
+        ipv4_address = Combine(ipv4_oct + ('.' + ipv4_oct*3)).setResultsName('ipv4_address')
```

Comment:
So there are two groups here; "comp_ipv4_address" and "ipv4_address". They match identical strings but the later is named (setResultsName). If they are named or not will give different results (having lots of named components means a deep nested dict is returned) when used to match, which is why I also have comp_ipv4_address, "comp" is a short for component and it is meant to be used as a component in a larger matching, in this case it is used for vrf_rt, which you can see on line 106. I don't care about  VRF RT "1.2.3.4:789" being built out of an ipv4_address : number, I just want to see it as one.

"ipv4_address" isn't actually used at all. I wanted to use it but had some issues and decided it was easier to leave it out for now. We still do interpret "1.2.3.0/24" as an IPv4 prefix, but pyparsing thinks it is a text and it is first on line 471 (in the PrefixSmartParsing) that we do parsing of this text string to identify IPv4/IPv6 prefixes. As you might notice, it is the code from our previous smart_parsing function that has been moved over.